design/Merkle.md

```diff
@@ -0,0 +1,226 @@
+
+- SHA1 / SHA256 / SHA3: `S` is an arbitrary sequence of bits
+- some less-conforming implementation of these could take a sequence of bytes instead
+- Poseidon: `S` is a sequence of finite field elements
+- Poseidon compression function: at most `t-1` field elements (in our case `t=3`, so 
+  that's two field elements)
+- A naive Merkle tree implementation could for example accept only a power-of-two 
+  sized sequence of `T`
+
+Notation: Let's denote a sequence of `T` by `[T]`.
+
+### Merkle tree API
+
+We usually need at least two types of Merkle tree APIs:
+
+- one which takes a sequence `S = [T]` of length `n` as input, and produces an 
+  output (Merkle root) of type `T`
+- and one which takes a sequence of bytes (or even bits, but in practice we probably 
+  only need bytes): `S = [byte]`
+
+We can decompose the latter into the composition of a function 
+`deserialize : [byte] -> [T]` and the former.
+
+### Naive Merkle tree implementation
+
+A straightforward implementation of a binary Merkle tree `merkleRoot : [T] -> T` 
+could be for example:
+
+- if the input has length 1, it's the root
+- if the input has even length `2*k`, group it into pairs, apply a 
+  `compress : (T,T) -> T` compression function, producing the next layer of size `k`
+- if the input has odd length `2*k+1`, pad it with an extra element `dummy` of 
+  type `T`, then apply the procedure for even length, producing the next layer of size `k+1`
+
+The compression function could be implemented in several ways:
+
+- when `S` and `T` are just sequences of bits or bytes (as in the case of classical hash
+  functions like SHA256), we can just concatenate the two leaves of the node and apply the
+  hash: `compress(x,y) := H(x|y)`
+- in case of hash functions based on the sponge construction (like Poseidon or Keccak/SHA3), 
+  we can just fill the "capacity part" of the state with a constant (say 0), the "absorbing 
+  part" of the state with the two inputs, apply the permutation, and extract a single `T` 
+
+### Attacks
+
+When implemented without enough care (like the above naive algorithm), there are several 
+possible attacks producing hash collisions or second preimages:
+
+1. The root of particular any layer is the same as the root of the input
+2. The root of `[x_0,x_1,...,x_(2*k)]` (length is `n=2*k+1` is the same as the root of 
+   `[x_0,x_1,...,x_(2*k),dummy]` (length is `n=2*k+2`)
+3. when using bytes as the input, already `deserialize` can have similar collision attacks
+4. The root of a singleton sequence is itself
+
+Traditional (linear) hash functions usually solve the analogous problems by clever padding.
+
+### Domain separation
+
+It's a good practice in general to ensure that different constructions using the same 
+underlying hash functions will never produce the same output. This is called "domain separation",
+and it's a bit similar to _multihash_; however instead of adding extra bits of information to a hash
```

Comment:
This is a bit of a misleading statement, `multihash` isn't a cryptographic construct at all, it is a meta-protocol for automatic negotiation in communication protocols.

Reply:
What I meant (but it could be probably phrased better), is that since in multihash you have some extra information telling you _which_ hash you use, it can never happen that you confuse two differently constructed hashes. 

Domain separation achieves the same (probabilistically), by putting this information into the hash. So for example you cannot confuse sponges with different rates (it could only happen for the empty input anyway, but still).

Reply:
> What I meant (but it could be probably phrased better), is that since in multihash you have some extra information telling you _which_ hash you use, it can never happen that you confuse two differently constructed hashes.
> 
> Domain separation achieves the same (probabilistically), by putting this information into the hash. So for example you cannot confuse sponges with different rates (it could only happen for the empty input anyway, but still).

Right, but this are different things in the sense that, the domain separator is never part of the resulting hash, where `multihash`, exist specifically to identify the type of hash being transmitted or stored and it is always packed with the hash and can be interpreted later on. 

I mean, I understand the analogy you're trying to make, but I think it adds confusion because the two constructs serve completely different purposes. No biggie, but I had to stop and think what you meant in that section.

Reply:
I changed the language to make it hopefully more clear that this is only a very loose analogy (i still think it's useful to mention it).